### PR TITLE
dk_use_iodata_for_datetime_formatting

### PR DESCRIPTION
**Pull request by dkuku**
The benchmark with go made me think if we can get any speed gains in synthetic tests :D
This pr changes datetime formatting functions:
around 55% of memory usage and 3x speedup when serializing with optional 4x when using iodata ;)

It also adds new functions `*_to_iodata` that can be used by libraries that operate on iodata like ecto or json
```elixir
defmodule CalendarStringBench do
  def run do
    datetime = ~U[2024-12-30 12:21:45.000123Z]

    Benchee.run(
      %{
        "Elixir.DateTime.to_string" => fn -> DateTime.to_string(datetime) end,
        "Calendar.ISO.datetime_to_string" => fn ->
          Calendar.ISO.naive_datetime_to_string(
            datetime.year,
            datetime.month,
            datetime.day,
            datetime.hour,
            datetime.minute,
            datetime.second,
            datetime.microsecond
          )
        end,
        "Calendar.IOISO.datetime_to_string" => fn ->
          Calendar.IOISO.naive_datetime_to_string(
            datetime.year,
            datetime.month,
            datetime.day,
            datetime.hour,
            datetime.minute,
            datetime.second,
            datetime.microsecond
          )
        end,
        "Calendar.IOISO.datetime_to_iodata" => fn ->
          Calendar.IOISO.naive_datetime_to_iodata(
            datetime.year,
            datetime.month,
            datetime.day,
            datetime.hour,
            datetime.minute,
            datetime.second,
            datetime.microsecond
          )
        end
      },
      time: 10,
      memory_time: 2,
      formatters: [
        #  {Benchee.Formatters.HTML, file: "bench/output/calendar_string.html"},
        Benchee.Formatters.Console
      ]
    )
  end
end

# Run the benchmark
CalendarStringBench.run()
```
result
```
Benchmarking Calendar.IOISO.datetime_to_iodata ...                                                                                            
Benchmarking Calendar.IOISO.datetime_to_string ...                                                                                            
Benchmarking Calendar.ISO.datetime_to_string ...                                                                                              
Benchmarking Elixir.DateTime.to_string ...                                                                                                    
Calculating statistics...                                                                                                                     
Formatting results...                                                                                                                         
                                                                                                                                              
Name                                        ips        average  deviation         median         99th %                                       
Calendar.IOISO.datetime_to_iodata        3.80 M      263.30 ns ±12989.04%         201 ns         350 ns                                       
Calendar.IOISO.datetime_to_string        2.86 M      350.05 ns  ±6363.30%         290 ns         471 ns                                       
Calendar.ISO.datetime_to_string          1.10 M      905.92 ns  ±3587.70%         651 ns        1352 ns                                       
Elixir.DateTime.to_string                1.04 M      958.94 ns  ±2794.13%         701 ns        1673 ns                                       
                                                                                                                                              
Comparison:                                                                                                                                   
Calendar.IOISO.datetime_to_iodata        3.80 M                                                                                               
Calendar.IOISO.datetime_to_string        2.86 M - 1.33x slower +86.75 ns
Calendar.ISO.datetime_to_string          1.10 M - 3.44x slower +642.62 ns
Elixir.DateTime.to_string                1.04 M - 3.64x slower +695.65 ns

Memory usage statistics:

Name                                 Memory usage                      
Calendar.IOISO.datetime_to_iodata           480 B                      
Calendar.IOISO.datetime_to_string           528 B - 1.10x memory usage +48 B
Calendar.ISO.datetime_to_string             896 B - 1.87x memory usage +416 B
Elixir.DateTime.to_string                   896 B - 1.87x memory usage +416 B
```

### CODE REVIEW

File: lib/elixir/lib/calendar/iso.ex [hunk 1]
```diff
@@ -1228,24 +1239,27 @@ defmodule Calendar.ISO do
...
   end
 
-  defp time_to_string_guarded(hour, minute, second, {microsecond, precision}, format) do
-    time_to_string_format(hour, minute, second, format) <>
-      "." <> (microsecond |> zero_pad(6) |> binary_part(0, precision))
+  defp time_to_iodata_guarded(hour, minute, second, {microsecond, precision}, format) do
+    [
+      time_to_iodata_format(hour, minute, second, format),
+      ".",
+      microsecond |> zero_pad(6) |> IO.iodata_to_binary() |> binary_part(0, precision)
```
Comment (dkuku) [inline lib/elixir/lib/calendar/iso.ex:1253]: there may be a better way for this ?

Comment (josevalim) [reply lib/elixir/lib/calendar/iso.ex:1253]: Pass the precision to zero pad and do it in one pass?

File: lib/elixir/lib/calendar/iso.ex [hunk 2]
```diff
@@ -1490,11 +1579,20 @@ defmodule Calendar.ISO do
...
-    :binary.copy("0", max(count - byte_size(num), 0)) <> num
+
+    case max(count - byte_size(num), 0) do
+      0 -> num
+      1 -> ["0", num]
+      2 -> ["00", num]
+      3 -> ["000", num]
+      4 -> ["0000", num]
+      5 -> ["00000", num]
+      6 -> ["000000", num]
```
Comment (josevalim) [inline lib/elixir/lib/calendar/iso.ex:1590]: I don't think we would ever reach this, since byte_size is never 0. We should also check that `count <= 6`.

**Comment by josevalim** [comment]
Yes, please do complete this pull request and expose these functions. Then please change JSON.Encoder to use the iodata variants for Calendar.ISO. :) Thank you!

File: lib/elixir/lib/calendar/iso.ex [hunk 3]
```diff
@@ -1273,18 +1287,27 @@ defmodule Calendar.ISO do
...
       when is_integer(year) and is_integer(month) and is_integer(day) and
              format in [:basic, :extended] do
-    date_to_string_guarded(year, month, day, format)
+    date_to_iodata_guarded(year, month, day, format)
   end
 
-  defp date_to_string_guarded(year, month, day, :extended) do
-    zero_pad(year, 4) <> "-" <> zero_pad(month, 2) <> "-" <> zero_pad(day, 2)
+  defp date_to_iodata_guarded(year, month, day, :extended) do
+    [zero_pad(year, 4), "-", zero_pad(month, 2), "-", zero_pad(day, 2)]
```
Comment (whatyouhide) [inline lib/elixir/lib/calendar/iso.ex:1302]:
```suggestion
    [zero_pad(year, 4), ?-, zero_pad(month, 2), ?-, zero_pad(day, 2)]
```

This should save a small amount of mem, but hey, why not!

File: lib/elixir/lib/calendar/iso.ex [hunk 4]
```diff
@@ -1490,11 +1579,20 @@ defmodule Calendar.ISO do
...
+      3 -> ["000", num]
+      4 -> ["0000", num]
+      5 -> ["00000", num]
+      6 -> ["000000", num]
+    end
   end
 
   defp zero_pad(val, count) do
-    "-" <> zero_pad(-val, count)
+    ["-", zero_pad(-val, count)]
```
Comment (whatyouhide) [inline lib/elixir/lib/calendar/iso.ex:1595]: Same here

```suggestion
    [?-, zero_pad(-val, count)]
```

File: lib/elixir/lib/calendar/iso.ex [hunk 5]
```diff
@@ -1228,24 +1239,27 @@ defmodule Calendar.ISO do
...
+      time_to_iodata_format(hour, minute, second, format),
+      ".",
+      microsecond |> zero_pad(6) |> IO.iodata_to_binary() |> binary_part(0, precision)
+    ]
   end
 
-  defp time_to_string_format(hour, minute, second, :extended) do
-    zero_pad(hour, 2) <> ":" <> zero_pad(minute, 2) <> ":" <> zero_pad(second, 2)
+  defp time_to_iodata_format(hour, minute, second, :extended) do
+    [zero_pad(hour, 2), ":", zero_pad(minute, 2), ":", zero_pad(second, 2)]
```
Comment (whatyouhide) [inline lib/elixir/lib/calendar/iso.ex:1258]: Same here (as below):

```suggestion
    [zero_pad(hour, 2), ?:, zero_pad(minute, 2), ?:, zero_pad(second, 2)]
```

**Comment by dkuku** [comment]
Do we really use other precisions than 0,3 and 6? 
```
Name                                        ips        average  deviation         median         99th %
Calendar.IOISO.datetime_to_iodata        4.75 M      210.37 ns ±14119.07%         170 ns         281 ns
Calendar.IOISO.datetime_to_string        3.20 M      312.52 ns  ±5866.09%         271 ns         420 ns
Calendar.ISO.datetime_to_string          1.12 M      893.28 ns  ±3302.84%         671 ns        1403 ns
Elixir.DateTime.to_string                1.09 M      915.88 ns  ±2761.10%         701 ns        1353 ns

Comparison: 
Calendar.IOISO.datetime_to_iodata        4.75 M
Calendar.IOISO.datetime_to_string        3.20 M - 1.49x slower +102.15 ns
Calendar.ISO.datetime_to_string          1.12 M - 4.25x slower +682.91 ns
Elixir.DateTime.to_string                1.09 M - 4.35x slower +705.51 ns

Memory usage statistics:

Name                                 Memory usage
Calendar.IOISO.datetime_to_iodata           456 B
Calendar.IOISO.datetime_to_string           504 B - 1.11x memory usage +48 B
Calendar.ISO.datetime_to_string             896 B - 1.96x memory usage +440 B
Elixir.DateTime.to_string                   896 B - 1.96x memory usage +440 B
```

**Comment by josevalim** [comment]
Precision can be any number between 0 and 6, even though we commonly use 0, 3, and 6.

**Comment by dkuku** [comment]
@whatyouhide I updated it but according to benchee the memory usage does not change - it's still 456B
I have between 4.6M -5.1M ips depending on the run. 
When I change the all the zeros from "000" to [?0, ?0, ?0] the memory usage goes up to 488B
```
Name                                        ips        average  deviation         median         99th %                                       
Calendar.IOISO.datetime_to_iodata        5.09 M      196.35 ns ±16231.76%         160 ns         270 ns                                       
                                                                                                                                              
Memory usage statistics:                                                                                                                      
                                                                                                                                              
Name                                 Memory usage                                                                                             
Calendar.IOISO.datetime_to_iodata           456 B                                                                                             
                                                                                                                                              
**All measurements for memory usage were the same**     
```

**Comment by dkuku** [comment]
> Yes, please do complete this pull request and expose these functions. Then please change JSON.Encoder to use the iodata variants for Calendar.ISO. :) Thank you!

@josevalim  This is currently implemented as:
```
defimpl JSON.Encoder, for: [Date, Time, NaiveDateTime, DateTime, Duration] do
  def encode(value, _encoder) do
    [?", @for.to_iso8601(value), ?"]
  end
end
```
I would have to create something like `to_iso8601_iodata` ?

**Comment by josevalim** [comment]
No need. That’s to_string for the iso calendar!

**Comment by josevalim** [comment]
All of the literals point to a constant, either ?0 or “0” take no extra runtime memory. The reason ?0, ?0 takes more than “00” is because it takes two entries in the list. Overall, a single character should be an integer, multiple should be a binary.

File: lib/elixir/lib/calendar/iso.ex [hunk 6]
```diff
@@ -1485,16 +1631,24 @@ defmodule Calendar.ISO do
...
+      2 -> ["00", num]
+      3 -> ["000", num]
+      4 -> ["0000", num]
+      5 -> ["00000", num]
+    end
   end
 
   defp zero_pad(val, count) do
-    "-" <> zero_pad(-val, count)
+    [?-, zero_pad(-val, count)]
```
[the remark this answers is not in the document]
Comment (dkuku) [reply lib/elixir/lib/calendar/iso.ex]: @sabiwara it makes a difference in the memory usage:
```
Name                                        ips        average  deviation         median         99th %
Calendar.IOISO.datetime_to_iodata        4.89 M      204.42 ns ±19386.20%         160 ns         271 ns

Memory usage statistics:

Name                                 Memory usage
Calendar.IOISO.datetime_to_iodata           376 B

**All measurements for memory usage were the same**
```

File: lib/elixir/lib/calendar/iso.ex [hunk 7]
```diff
@@ -1261,15 +1269,16 @@ defmodule Calendar.ISO do
...
   end
 
-  defp microseconds_to_iodata(_microsecond, 0), do: []
-  defp microseconds_to_iodata(microsecond, 6), do: zero_pad(microsecond, 6)
+  @doc false
+  def microseconds_to_iodata(_microsecond, 0), do: []
+  def microseconds_to_iodata(microsecond, 6), do: zero_pad(microsecond, 6)
 
-  defp microseconds_to_iodata(microsecond, precision) do
+  def microseconds_to_iodata(microsecond, precision) do
```
Comment (dkuku) [inline lib/elixir/lib/calendar/iso.ex:1281]: I need it in the next pr

File: lib/elixir/lib/calendar/iso.ex [hunk 8]
```diff
@@ -1273,18 +1322,35 @@ defmodule Calendar.ISO do
...
   @impl true
-  def date_to_string(year, month, day, format \\ :extended)
+  def date_to_string(year, month, day, format \\ :extended) do
+    date_to_iodata(year, month, day, format)
+    |> IO.iodata_to_binary()
+  end
+
+  @doc """
+  Converts the given date into a iodata.
+  Look at date_to_string/4 for more information
```
Comment (sabiwara) [inline lib/elixir/lib/calendar/iso.ex:1332]:
```suggestion
  Converts the given date into an iodata.
  
  See `date_to_string/4` for more information.
```

File: lib/elixir/lib/calendar/iso.ex [hunk 9]
```diff
@@ -1220,6 +1220,37 @@ defmodule Calendar.ISO do
...
+        format \\ :extended
+      ) do
+    time_to_iodata(hour, minute, second, microsecond, format)
+    |> IO.iodata_to_binary()
+  end
+
+  @doc """
+  Converts the given time into a iodata.
+
+  Look at time_to_string/5 for more information
```
Comment (sabiwara) [inline lib/elixir/lib/calendar/iso.ex:1236]:
```suggestion
  Converts the given time into an iodata.

  See `time_to_string/5` for more information.
```

File: lib/elixir/lib/calendar/iso.ex [hunk 10]
```diff
@@ -1327,8 +1393,60 @@ defmodule Calendar.ISO do
...
+      second,
+      microsecond,
+      format
+    )
+    |> IO.iodata_to_binary()
+  end
+
+  @doc """
+  Converts the given naive_datetime into a iodata.
+  Look at naive_datetime_to_iodata/8 for more information
```
Comment (sabiwara) [inline lib/elixir/lib/calendar/iso.ex:1411]:
```suggestion
  Converts the given naive_datetime into an iodata.
  
  See `naive_datetime_to_iodata/8` for more information.
```

File: lib/elixir/lib/calendar/iso.ex [hunk 11]
```diff
@@ -1394,20 +1512,88 @@ defmodule Calendar.ISO do
...
+      utc_offset,
+      std_offset,
+      format
+    )
+    |> IO.iodata_to_binary()
+  end
+
+  @doc """
+  Converts the given datetime into a iodata.
+  Look at datetime_to_iodata/12 for more information
```
Comment (sabiwara) [inline lib/elixir/lib/calendar/iso.ex:1535]:
```suggestion
  Converts the given datetime into an iodata.
  
  See `datetime_to_iodata/12` for more information.
```

**Review by sabiwara — APPROVED** [review]
Amazing, thanks @dkuku 💜

**Comment by sabiwara** [comment]
> but according to benchee the memory usage does not change

> All of the literals point to a constant, either ?0 or “0” take no extra runtime memory. <...> Overall, a single character should be an integer, multiple should be a binary.

In this case rather than the memory of using chars (except if we copy the deep iolist), the most noticeable benefit seems to be speed downstream when actually using the iodata e.g. to create a binary. Sharing this TIL I discovered when benching, cf https://github.com/erlang/otp/pull/9251.

File: lib/elixir/lib/calendar/iso.ex [hunk 12]
```diff
@@ -1220,6 +1220,37 @@ defmodule Calendar.ISO do
...
+        format \\ :extended
+      ) do
+    time_to_iodata(hour, minute, second, microsecond, format)
+    |> IO.iodata_to_binary()
+  end
+
+  @doc """
+  Converts the given time into a iodata.
+
+  See time_to_string/5 for more information
```
Comment (sabiwara) [inline lib/elixir/lib/calendar/iso.ex:1236]:
```suggestion
  See `time_to_string/5` for more information.
```

File: lib/elixir/lib/calendar/iso.ex [hunk 13]
```diff
@@ -1273,18 +1322,36 @@ defmodule Calendar.ISO do
...
-  def date_to_string(year, month, day, format \\ :extended)
+  def date_to_string(year, month, day, format \\ :extended) do
+    date_to_iodata(year, month, day, format)
+    |> IO.iodata_to_binary()
+  end
+
+  @doc """
+  Converts the given date into a iodata.
+
+  See date_to_string/4 for more information
```
Comment (sabiwara) [inline lib/elixir/lib/calendar/iso.ex:1333]:
```suggestion
  See `date_to_string/4` for more information.
```

File: lib/elixir/lib/calendar/iso.ex [hunk 14]
```diff
@@ -1327,8 +1394,61 @@ defmodule Calendar.ISO do
...
+      microsecond,
+      format
+    )
+    |> IO.iodata_to_binary()
+  end
+
+  @doc """
+  Converts the given naive_datetime into a iodata.
+
+  See naive_datetime_to_iodata/8 for more information
```
Comment (sabiwara) [inline lib/elixir/lib/calendar/iso.ex:1413]:
```suggestion
  See `naive_datetime_to_iodata/8` for more information.
```

File: lib/elixir/lib/calendar/iso.ex [hunk 15]
```diff
@@ -1394,20 +1514,89 @@ defmodule Calendar.ISO do
...
+      std_offset,
+      format
+    )
+    |> IO.iodata_to_binary()
+  end
+
+  @doc """
+  Converts the given datetime into a iodata.
+
+  See datetime_to_iodata/12 for more information
```
Comment (sabiwara) [inline lib/elixir/lib/calendar/iso.ex:1538]:
```suggestion
  See `datetime_to_iodata/12` for more information.
```

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Comment by dkuku** [comment]
There is one small micro optimization we can make here that is still visible: a special case for zero_pad with 2 elements, which is called quite often. The math is really simple and doesn't require calling `Integer.to_string`.

```elixir
  defp zero_pad(val, 2) when val >= 0 and val < 10 do
    [?0, val + ?0]
  end

  defp zero_pad(val, 2) when val >= 10 and val < 100 do
    tens = div(val, 10)
    [tens + ?0, val - tens * 10 + ?0]
  end
```
Benchee + fprof results after applying this change:
```
Name                                        ips        average  deviation         median         99th %
Calendar.IOISO.datetime_to_string        4.28 M      233.68 ns  ±5211.52%         201 ns         321 ns
Calendar.ISO.datetime_to_string          3.70 M      270.57 ns  ±5605.04%         231 ns         361 ns

Comparison: 
Calendar.IOISO.datetime_to_string        4.28 M
Calendar.ISO.datetime_to_string          3.70 M - 1.16x slower +36.89 ns

Extended statistics: 

Name                                      minimum        maximum    sample size                     mode
Calendar.IOISO.datetime_to_string          160 ns    16775502 ns         5.26 M                   200 ns
Calendar.ISO.datetime_to_string            200 ns    17099465 ns         4.84 M                   230 ns

Memory usage statistics:

Name                                 Memory usage
Calendar.IOISO.datetime_to_string           464 B
Calendar.ISO.datetime_to_string             424 B - 0.91x memory usage -40 B

**All measurements for memory usage were the same**

Profiling Calendar.IOISO.datetime_to_string with eprof...

Profile results of #PID<0.1123558.0>
#                                           CALLS     % TIME µS/CALL
Total                                          21 100.0    7    0.33
anonymous fn/0 in CalendarStringBench.run/0     1  0.00    0    0.00
Calendar.IOISO.time_to_iodata_guarded/5         1  0.00    0    0.00
Calendar.IOISO.time_to_iodata_format/4          1  0.00    0    0.00
Calendar.IOISO.time_to_iodata/5                 1  0.00    0    0.00
Calendar.IOISO.naive_datetime_to_string/8       1  0.00    0    0.00
Calendar.IOISO.naive_datetime_to_string/7       1  0.00    0    0.00
Calendar.IOISO.naive_datetime_to_iodata/8       1  0.00    0    0.00
Calendar.IOISO.microseconds_to_iodata/2         1  0.00    0    0.00
Calendar.IOISO.date_to_iodata_guarded/4         1  0.00    0    0.00
Calendar.IOISO.date_to_iodata/4                 1  0.00    0    0.00
:erlang.integer_to_binary/1                     2 14.29    1    0.50
Calendar.IOISO.zero_pad/2                       7 14.29    1    0.14
:erlang.iolist_to_binary/1                      1 28.57    2    2.00
:erlang.apply/2                                 1 42.86    3    3.00

Profile done over 14 matching functions

Profiling Calendar.ISO.datetime_to_string with eprof...

Profile results of #PID<0.1123560.0>
#                                           CALLS     % TIME µS/CALL
Total                                          26 100.0    9    0.35
Calendar.ISO.time_to_iodata_guarded/5           1  0.00    0    0.00
Calendar.ISO.time_to_iodata_format/4            1  0.00    0    0.00
Calendar.ISO.time_to_iodata/5                   1  0.00    0    0.00
Calendar.ISO.naive_datetime_to_string/8         1  0.00    0    0.00
Calendar.ISO.naive_datetime_to_string/7         1  0.00    0    0.00
Calendar.ISO.naive_datetime_to_iodata/8         1  0.00    0    0.00
Calendar.ISO.microseconds_to_iodata/2           1  0.00    0    0.00
Calendar.ISO.date_to_iodata_guarded/4           1  0.00    0    0.00
Calendar.ISO.date_to_iodata/4                   1  0.00    0    0.00
anonymous fn/0 in CalendarStringBench.run/0     1 11.11    1    1.00
:erlang.apply/2                                 1 22.22    2    2.00
:erlang.iolist_to_binary/1                      1 22.22    2    2.00
:erlang.integer_to_binary/1                     7 22.22    2    0.29
Calendar.ISO.zero_pad/2                         7 22.22    2    0.29

Profile done over 14 matching functions
```

TIL `[1, 2]` is iodata but `[1 | 2]` is not
```elixir
iex(1)> IO.iodata_to_binary([1, 2])
<<1, 2>>
iex(2)> IO.iodata_to_binary([1 | 2])
** (ArgumentError) errors were found at the given arguments:

  * 1st argument: not an iodata term

    :erlang.iolist_to_binary([1 | 2])
    iex:2: (file)
```

**Comment by michalmuskala** [comment]
One thing to be careful with iodata is when you keep it around for a long time - traversing iodata during GC is very, very expensive compared to binaries.
We've hit that in Jason before - where eagerly converting stuff to binaries, rather than keeping as iodata was beneficial.

The benchmarks here are only dealing with allocated memory, not retained, so they don't capture the issue. The way to capture the issue would be to encode a list of dates and return a list of encoded values, rather than doing it one by one.

**Comment by dkuku** [comment]
Good point but without a proper benchmark it's hard to compare.
The intermediate short strings that are concatenated also need to be GCed.

**Comment by michalmuskala** [comment]
If the code is written to take advantage of mutable binaries, it's possible to avoid that - there will be just one binary buffer that gets appended to

**Comment by dkuku** [comment]
You made me curious. 
I [prepared a branch](https://github.com/dkuku/elixir/pull/1) and run a benchmark.
The speed is the same as with binary concatenation, but the memory usage reported by benchee it over 3kb (benchmark result in the pr desc).
It's probably because I'm appending to the buffer.

**Pull request merged**